### PR TITLE
Bug: delete all policies when management + hub hosted mode

### PR DESCRIPTION
**Pull request by yiraeChristineKim**
Bug: delete all policies when management + hub hosted mode

description:
When hcp clusters are created on the MCV2, policies are rolled out well. When deleting the hcp cluster, we see that all the ACM policies on the ACM hub are also deleted.

ref: 97d54e1e64afa56b22327b1c97b04d31cc2db7b4

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -68,6 +68,15 @@ rules:
...
   - update
+- apiGroups:
+  - apiextensions.k8s.io
+  resources:
+  - customresourcedefinitions
+  verbs:
+  - get
+  - list
+  - patch
+  - update
```
Comment (yiraeChristineKim) [inline config/rbac/role.yaml:79]: I will open pr on multiclusterhub

Comment (yiraeChristineKim) [reply config/rbac/role.yaml:79]: https://github.com/stolostron/multiclusterhub-operator/pull/1069

File: config/manager/manager.yaml [hunk 2]
```diff
@@ -44,6 +44,8 @@ spec:
           value: registry.redhat.io/openshift4/ose-kube-rbac-proxy:v4.10
         - name: GOVERNANCE_POLICY_FRAMEWORK_ADDON_IMAGE
           value: quay.io/open-cluster-management/governance-policy-framework-addon:latest
+        - name: HOSTING_CLUSTER_NAME
```
Comment (mprahl) [inline config/manager/manager.yaml:47]: How do you expect this to get populated? The hosting cluster can be different for different hosted clusters.

Comment (yiraeChristineKim) [reply config/manager/manager.yaml:47]: This env is for our kind test so the default is` local-cluser`

Comment (yiraeChristineKim) [reply config/manager/manager.yaml:47]: so in Prod this going to be empty

**Comment by mprahl** [comment]
@yiraeChristineKim, the ManagedClusterAddOn objects record what the hosting cluster name is. You just need to check if it's equal to `local-cluster`.

**Comment by yiraeChristineKim** [comment]
> @yiraeChristineKim, the ManagedClusterAddOn objects record what the hosting cluster name is. You just need to check if it's equal to `local-cluster`.

cluster1 is for our e2e test.

**Comment by mprahl** [comment]
@yiraeChristineKim this is what I was thinking but I didn't test it:

```diff
diff --git a/pkg/addon/policyframework/agent_addon.go b/pkg/addon/policyframework/agent_addon.go
index 3ac1a19..a56ac14 100644
--- a/pkg/addon/policyframework/agent_addon.go
+++ b/pkg/addon/policyframework/agent_addon.go
@@ -74,8 +74,13 @@ func getValues(cluster *clusterv1.ManagedCluster,
                        PkgLogLevel: -1,
                },
        }
-       // special case for local-cluster
-       if cluster.Name == "local-cluster" {
+
+       annotations := addon.GetAnnotations()
+
+       hostingClusterName := annotations[addonapiv1alpha1.HostingClusterNameAnnotationKey]
+
+       // special case when running on the Hub
+       if cluster.Name == "local-cluster" || hostingClusterName == "local-cluster" {
                userValues.OnMulticlusterHub = true
        }
 
@@ -87,8 +92,6 @@ func getValues(cluster *clusterv1.ManagedCluster,
                }
        }
 
-       annotations := addon.GetAnnotations()
-
        if val, ok := annotations["addon.open-cluster-management.io/on-multicluster-hub"]; ok {
                if strings.EqualFold(val, "true") {
                        userValues.OnMulticlusterHub = true
diff --git a/pkg/addon/policyframework/manifests/managedclusterchart/templates/cleanup_pod.yaml b/pkg/addon/policyframework/manifests/managedclusterchart/templates/cleanup_pod.yaml
index 51af153..5516887 100644
--- a/pkg/addon/policyframework/manifests/managedclusterchart/templates/cleanup_pod.yaml
+++ b/pkg/addon/policyframework/manifests/managedclusterchart/templates/cleanup_pod.yaml
@@ -1,6 +1,6 @@
 # Copyright Contributors to the Open Cluster Management project
 
-{{- if (not .Values.onMulticlusterHub) }}
+{{- if or (not .Values.onMulticlusterHub) (eq .Values.installMode "Hosted") }}
 apiVersion: v1
 kind: Pod
 metadata:
diff --git a/pkg/addon/policyframework/manifests/managedclusterchart/templates/deployment.yaml b/pkg/addon/policyframework/manifests/managedclusterchart/templates/deployment.yaml
index 55be0f4..6c2fb3d 100644
--- a/pkg/addon/policyframework/manifests/managedclusterchart/templates/deployment.yaml
+++ b/pkg/addon/policyframework/manifests/managedclusterchart/templates/deployment.yaml
@@ -72,7 +72,7 @@ spec:
           - --log-encoder={{ .Values.args.logEncoder }}
           - --log-level={{ .Values.args.logLevel }}
           - --v={{ .Values.args.pkgLogLevel }}
-          {{- if .Values.onMulticlusterHub }}
+          {{- if or (not .Values.onMulticlusterHub) (eq .Values.installMode "Hosted") }}
           - --disable-spec-sync=true
           {{- end }}
           {{- if eq .Values.installMode "Hosted" }}
```

**Comment by yiraeChristineKim** [comment]
> ```diff
> installMode
> ```

When management cluster and hubare are separated, Is  .Values.installMode same as "Hosted"

**Comment by yiraeChristineKim** [comment]
@mprahl 
Also` cluster.Name == "local-cluster" ` this is not the case "management cluster + hub"
` cluster(managedcluster).Name == "local-cluster" means one of the hosted clusters is a hub cluster. I think it is another case. Is my understanding is wrong? I thought management cluster == hosting cluster

**Comment by mprahl** [comment]
> > ```diff
> > installMode
> > ```
> 
> When management cluster and hubare are separated, Is .Values.installMode same as "Hosted"

installMode is always `Hosted` if the addon is deployed in hosted mode regardless of the hosting cluster.

**Comment by mprahl** [comment]
> @mprahl Also`cluster.Name == "local-cluster"` this is not the case "management cluster + hub" ` cluster(managedcluster).Name == "local-cluster" means one of the hosted clusters is a hub cluster. It is another case. Is my understanding wrong? I thought management cluster == hosting cluster

True, but in the tests `onMulticlusterHub` is explictly set:
https://github.com/open-cluster-management-io/governance-policy-addon-controller/blob/cbcabf58f51a94b780b928578d1b20220f8493b4/test/resources/framework_hub_annotation_addon_cr.yaml#L5

**Comment by yiraeChristineKim** [comment]
@mprahl  I don't understand .why `Values.installMode "Hosted"` added on clean-up pod and what does `--disable-spec-sync=true` do? Can you explain to me?

**Comment by mprahl** [comment]
> @mprahl I don't understand .why `Values.installMode "Hosted"` added on clean-up pod and what does `--disable-spec-sync=true` do? Can you explain to me?

Sorry, I had a mistake in the example diff. I just updated it.

`--disable-spec-sync` is required when running on the hub not in hosted mode. This is because the managed cluster namespace on the hub has its policies placed by the Policy Propagator and the governance-policy-framework addon uses that same namespace when it's self-managed Hub. There is no need to copy those policies.

In theory, if a hosted cluster has its addons running on the ACM hub and the same managed cluster namespace is used, we could also disable the spec sync in that case but I don't think we should support that unless it's explicitly requested.

**Comment by yiraeChristineKim** [comment]
@mprahl  If we add `+{{- if and (not .Values.onMulticlusterHub) (ne .Values.installMode "Hosted") }}` on clean up pod...  it never falls into   line 28 on cleanup_pod.yaml    ```
  {{- if eq .Values.installMode "Hosted" }}
        - --policy-namespace={{ .Release.Namespace }}
```

also  does this  line 14 on cleanup_pod.yaml   `addon.open-cluster-management.io/hosted-manifest-location: hosting ` has meaning?

**Comment by mprahl** [comment]
> @mprahl If we add `+{{- if and (not .Values.onMulticlusterHub) (ne .Values.installMode "Hosted") }}` on clean up pod... it never falls into line 28 on cleanup_pod.yaml ``` {{- if eq .Values.installMode "Hosted" }} - --policy-namespace={{ .Release.Namespace }}
> 
> ```
> 
> also  does this  line 14 on cleanup_pod.yaml   `addon.open-cluster-management.io/hosted-manifest-location: hosting ` has meaning? 
> ```

Sorry about that, it should be `{{- if or (not .Values.onMulticlusterHub) (eq .Values.installMode "Hosted") }}` because we only want to not have the clean up pod if it's a self-managed hub because the Policy Propagator owns this namespace on the hub.

Yes, `addon.open-cluster-management.io/hosted-manifest-location: hosting` will determine where this gets deployed to in hosted mode. In this case, the hosting cluster.

**Comment by yiraeChristineKim** [comment]
@mprahl   self-managed hub means it is onMulticlusterHub, not hosted mode... is this correct? I thought onMulticlusterHub means acm

**Comment by mprahl** [comment]
> @mprahl self-managed hub means it is onMulticlusterHub, not hosted mode... is this correct? I thought onMulticlusterHub means acm

Yes, that is correct.

**Comment by yiraeChristineKim** [comment]
@mprahl How about our e2e test, hub cluster is always cluster1 not local-cluster

**Comment by mprahl** [comment]
> @mprahl How about our e2e test, hub cluster is always cluster1 not local-cluster

`onMulticlusterHub` can be explicitly set it's just a helper that if it's `local-cluster` then it's automatically set and the user doesn't need to provide additional information.

Please refer to my previous comment for the code line in the test:
https://github.com/open-cluster-management-io/governance-policy-addon-controller/pull/94#issuecomment-1673726227

File: test/e2e/case2_config_deployment_test.go [hunk 3]
```diff
@@ -30,13 +32,13 @@ func verifyConfigPolicyDeployment(
 	By(logPrefix + "checking the number of containers in the deployment")
 
 	deploy := GetWithTimeout(
-		client, gvrDeployment, case2DeploymentName, namespace, true, 60,
+		client, gvrDeployment, case2DeploymentName, namespace, true, 300,
```
Comment (mprahl) [inline test/e2e/case2_config_deployment_test.go:35]: Are these additional timeout increases still required?

**Comment by mprahl** [comment]
@yiraeChristineKim could you please change your commit title to be about what the fix is as opposed to what the bug is? For example, `Fix hosted mode when hosting cluster is a hub`.

File: test/e2e/case2_config_deployment_test.go (hunk 3 above)
Comment (yiraeChristineKim) [reply test/e2e/case2_config_deployment_test.go:35]: The tests are so flakey Can I try few more?

**Comment by mprahl** [comment]
> @yiraeChristineKim could you please change your commit title to be about what the fix is as opposed to what the bug is? For example, `Fix hosted mode when hosting cluster is a hub`.

@yiraeChristineKim could you please follow the commit message guidelines here?
https://github.com/open-cluster-management-io/community/blob/main/sig-policy/contribution-guidelines.md#commit-messages

**Review by mprahl — COMMENTED** [review]
This looks good to me and I'll approve it after the commit message guidelines are followed.

**Comment by mprahl** [comment]
> > @yiraeChristineKim could you please change your commit title to be about what the fix is as opposed to what the bug is? For example, `Fix hosted mode when hosting cluster is a hub`.
> 
> @yiraeChristineKim could you please follow the commit message guidelines here? https://github.com/open-cluster-management-io/community/blob/main/sig-policy/contribution-guidelines.md#commit-messages

@yiraeChristineKim specifically, could you please follow this?
https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#use-imperative-mood-in-your-commit-message-subject

It should start with a verb in the imperative mood. Also, please don't include `Fix:` as a prefix since it's implied that it's a fix based on the commit message title.

**Review by mprahl — APPROVED** [review]

**Pull request merged**